Fix StringBuilder.insert and String.substring [TG-2459]

romainbrenguier opened the pull request:
This fixes two small issues with these methods, and adds tests for them.

romainbrenguier committed:
Fix string constraints for substring

It was previously not clear what could happen when the argument of
substring were out of bounds. This is now clearly specified and avoids
getting formulas with array accesses at negative indexes in particular.

romainbrenguier committed:
Test for CProverString.insert

This function behaves slightly differently from StringBuilder.insert
since it never raise an exception.
This tests the cases where the index exceeds the limits.

romainbrenguier committed:
Tests for CProverString.substring

romainbrenguier committed:
Fix add_axioms_for_concat_substr

What was happening for start and end index out of bound wasn't clear and
could lead to empty index set because the strings where sometimes
indexed at negative indices.
The behavior is now similar to substring: the actual start index is
`start_index' = max(0, start_index)` (where start_index is the argument value) and the
actual end index is `max(min(end_index, s2.length), start_index')`.

LAJW commented on src/solvers/refinement/string_refinement_util.cpp:
```diff
@@ -322,18 +322,22 @@ std::vector<mp_integer> string_insertion_builtin_functiont::eval(
-  const std::size_t &end = args_value.size() > 2
-                             ? numeric_cast_v<std::size_t>(args_value[2])
-                             : input2_value.size();
+  const auto offset = std::max<BigInt::llong_t>(args_value[0].to_long(), 0);
```
I'd put this repeated "if less than zero then zero" into a static "diode" function. Only I'm not sure about the name.

peterschrammel replied:
Don't use `BigInt` directly, use `mp_integer` instead.

LAJW commented on src/solvers/refinement/string_constraint_generator_concat.cpp:
```diff
@@ -44,28 +38,26 @@ exprt string_constraint_generatort::add_axioms_for_concat_substr(
+  const exprt end1 = maximum(minimum(end_index, s2.length()), start1);
+  const plus_exprt res_length(s1.length(), minus_exprt(end1, start1));
+  const equal_exprt a1(res.length(), res_length);
   lemmas.push_back(a1);
```
You could group those `a1`, `a2`, ... into:
```c++
lemmas.push_back([&] {
  exprt lemma;
  ...
  return lemma;
}());
```
That way you wouldn't have to worry about numbering these things.

romainbrenguier replied:
Numbering the axioms is not really a problem because they refer to these numbers in the documentation, but scoping some of these variables is definitely good, because I've made several times mistake where a universal symbol was used in two different constraints, so I'm going to follow your suggestion and add a comment to say to which axiom this corresponds

LAJW reviewed (approved):
LGTM, you could make it a little prettier if you had spare time.

peterschrammel commented on src/solvers/refinement/string_refinement_util.cpp:
```diff
@@ -322,18 +322,22 @@ std::vector<mp_integer> string_insertion_builtin_functiont::eval(
+
+  const auto end = std::max<BigInt::llong_t>(
+    std::min<BigInt::llong_t>(
+      args_value.size() > 2 ? args_value[2].to_long()
```
I don't understand what the purpose of all these number conversions is. It would be much simpler to use `mp_integer` throughout.

romainbrenguier replied:
@peterschrammel To make that easier I added max and min functions to `mp_integer` in the first commit, can you have a look at that?

peterschrammel commented on src/solvers/refinement/string_constraint_generator_concat.cpp:
```diff
@@ -44,28 +38,26 @@ exprt string_constraint_generatort::add_axioms_for_concat_substr(
+  const exprt start1 = maximum(start_index, from_integer(0, index_type));
+  const exprt end1 = maximum(minimum(end_index, s2.length()), start1);
+  const plus_exprt res_length(s1.length(), minus_exprt(end1, start1));
+  const equal_exprt a1(res.length(), res_length);
```
(comment for commit message) What is happening to index out of bounds now? The commit message doesn't give much information.

romainbrenguier replied:
commit message updated

romainbrenguier committed:
Fix eval of insert to match string constraints

String constraints for insert are permisive with index that are negative
or exceed the length of the string so the eval method should do the
same, and in particular should not make an invariant fail.

tautschnig commented on src/solvers/refinement/string_refinement_util.cpp:
```diff
@@ -322,18 +322,20 @@ std::vector<mp_integer> string_insertion_builtin_functiont::eval(
-    input2_value.end() + end);
+    result.begin() + offset.to_ulong(),
+    input2_value.begin() + start.to_ulong(),
+    input2_value.begin() + end.to_ulong());
```
Use `integer2size_t` or `numeric_cast<std::size_t` in all of the above.

tautschnig commented on src/solvers/refinement/string_refinement_util.cpp:
```diff
@@ -322,18 +322,20 @@ std::vector<mp_integer> string_insertion_builtin_functiont::eval(
+  const auto end = max(
+    min(args_value.size() > 2 ? args_value[2] : mp_integer(input2_value.size()),
+        input2_value.size()),
+    0);
```
Can't we just use `std::min` and `std::max`? Or really rewrite the entire code?
````c++
std::size_t offset = 0, start = 0, end = 0;
if(args_value[0] > 0)
  offset = integer2size_t(args_value[0]);

if(args_value.size() > 1 && args_value[1] > 0)
  start = integer2size_t(args_value[1]);

if(args_value.size() > 2 && args_value[2] > 0)
  end = std::min(integer2size_t(args_value[2]), input_value.size());
````

tautschnig commented on src/util/mp_arith.h:
```diff
@@ -62,4 +62,18 @@ unsigned integer2unsigned(const mp_integer &);
+mp_integer max(const mp_integer &m, const T& n)
+{
+  return m >= n ? m : mp_integer(n);
+}
```
Let's just not do this. We don't have to reinvent every library function with just a tiny tweak to its API.

romainbrenguier committed:
Fix eval of concatenation builtin functions

The previous version was not handling correcly the case where end_index
was smaller than start_index.

romainbrenguier committed:
Tests for CProverString.append with start/end args

This test the CProverString.append function when start_index and
end_index arguments are proivided.

romainbrenguier commented:
@LAJW I added a test for append and included a fix to the eval method of concatenation (was necessary to activate the test), you may want to have a look at the two last commits.

@peterschrammel this is ready for re-review

LAJW commented:
Still looks decent.
